gtsam/navigation/BarometricFactor.cpp

```diff
@@ -0,0 +1,53 @@
+/* ----------------------------------------------------------------------------
+
+ * GTSAM Copyright 2010, Georgia Tech Research Corporation,
+ * Atlanta, Georgia 30332-0415
+ * All Rights Reserved
+ * Authors: Frank Dellaert, et al. (see THANKS for the full author list)
+
+ * See LICENSE for the license information
+
+ * -------------------------------------------------------------------------- */
+
+/**
+ *  @file   BarometricFactor.cpp
+ *  @author Peter Milani
+ *  @brief  Implementation file for Barometric factor
+ *  @date   December 16, 2021
+ **/
+
+#include "BarometricFactor.h"
+
+using namespace std;
+
+namespace gtsam {
+
+//***************************************************************************
+void BarometricFactor::print(const string& s,
+                             const KeyFormatter& keyFormatter) const {
+    cout << (s.empty() ? "" : s + " ") << "Barometric Factor on "
+         << keyFormatter(key1()) << "Barometric Bias on "
+         << keyFormatter(key2()) << "\n";
+
+    cout << "  Baro measurement: " << nT_ << "\n";
+    noiseModel_->print("  noise model: ");
+}
+
+//***************************************************************************
+bool BarometricFactor::equals(const NonlinearFactor& expected,
+                              double tol) const {
+    const This* e = dynamic_cast<const This*>(&expected);
+    return e != nullptr && Base::equals(*e, tol) &&
+           traits<double>::Equals(nT_, e->nT_, tol);
+}
+
+//***************************************************************************
+Vector BarometricFactor::evaluateError(const Pose3& p, const double& bias,
+                                       boost::optional<Matrix&> H,
+                                       boost::optional<Matrix&> H2) const {
+    if (H2) (*H2) = (Matrix(1, 1) << 1.0).finished();
+    if (H) (*H) = (Matrix(1, 6) << 0., 0., 0., 0., 0., 1.).finished();
```

Comment:
Respect the order :-) also, this is likely to be incorrect. The translation method will actually calculate the correct Jacobian for you.

Reply:
Thanks Frank, I've updated it to use the output of the Jacobian from the translation method. The jacobians have been something that I've found challenging for a while. I think its been mostly around the frames used as I typically have found it difficult to derive those given in all but the simplest of examples. The papers and tutorials I've read don't seem to go into depth on the subject. Are there any good tutorials on derivation of analytic jacobians for these type of systems?

It does seem that gtsam has a good system for dealing with this complexity, but I'd like to be better set up to handle these problems in future. 

@dellaert Thats all updated and for review.

Reply:
Great! I started to workflows for CI. As far as the math, take a look in math.pdf, in the docs folder.